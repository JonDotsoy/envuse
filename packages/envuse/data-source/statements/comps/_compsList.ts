export { Block } from "./Block";
export { CommentInline as Comment } from "./CommentInline";
export { CommentOperator } from "./CommentOperator";
export { CommentOperatorStatement } from "./CommentOperatorStatement";
export { Space } from "./Space";
export { SpaceNewLine } from "./SpaceNewLine";
export { StatementObject } from "./StatementObject";
export { SymbolEqual } from "./SymbolEqual";
export { Variable } from "./Variable";
export { VariableKey } from "./VariableKey";
export { VariableValue } from "./VariableValue";
