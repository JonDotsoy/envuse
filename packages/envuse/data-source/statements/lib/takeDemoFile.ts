import path from "path";
import fs from "fs";
import { range } from "./range";
import YAML from "yaml"

const getState = () => {
  const state = expect.getState()
  const demoPath = `${path.dirname(state.testPath)}/__demo__/${path.basename(state.testPath)}.yaml`;

  return {
    demoPath,
    jestExpectState: state,
  }
}

// const once = <T>(fn: () => T, cache?: T) => () => (cache = cache ?? fn(), cache)

const readFileCached = (() => {
  const cache = new Map<string, () => Buffer>();

  return (pathFile: string) => {
    const cached = cache.get(pathFile);

    if (cached) {
      return cached();
    }

    // Check file exists
    if (!fs.existsSync(pathFile)) {
      const initialBuff = Buffer.from([]);
      fs.writeFileSync(pathFile, initialBuff);
      // Log file correctly created
      console.log(`File ${pathFile} created`);
      return initialBuff
    }

    const file = fs.readFileSync(pathFile);
    cache.set(pathFile, () => file);

    return file;
  }
})();

const readDemoFile = (filePath: string) => {
  type BodyFile = string | {
    path: string
    body: Buffer
  }

  type Body = {
    [key: string]: BodyFile
  }

  const body: Body = YAML.parse(readFileCached(filePath).toString(), {
    customTags: [
      {
        identify: (tagName: string) => tagName === "file",
        tag: "!file",
        resolve(data: any, cst: any) {
          const pathFile = path.resolve(path.dirname(filePath), cst.strValue)

          if (!fs.existsSync(pathFile)) {
            throw new Error(`File ${pathFile} not found`);
          }

          return { path: pathFile, body: fs.readFileSync(pathFile) }
        }
      }
    ]
  })

  return body
}


const charAllow = Buffer.from([
  0x5f,
  0x2d,
  ...Array.from(range(0x61, 0x7a)),
  ...Array.from(range(0x41, 0x5a)),
  ...Array.from(range(0x30, 0x39)),
]);

export function takeDemoFile(sufixt?: string) {
  const { demoPath, jestExpectState } = getState();
  const relativeDemoPath = path.relative(process.cwd(), demoPath);
  const bodyDemofile = readDemoFile(demoPath);

  const testName = `${jestExpectState.currentTestName}${sufixt ?? ''}`;
  const demo = bodyDemofile?.[testName];

  if (typeof demo === "object" && demo.path) {

    return [demo.path, demo.body, bodyDemofile] as const

  }

  // Check exists demo
  if (!demo) {
    const err = new Error(`Cannot found demo ${JSON.stringify(testName)} on file ${relativeDemoPath}. Please create it on ${relativeDemoPath}`);
    Error.captureStackTrace(err, takeDemoFile);
    throw err;
  }

  // Validate body is a string
  if (typeof demo !== "string") {
    const err = new Error(`Demo ${JSON.stringify(testName)} on file ${relativeDemoPath} must be a string.`);
    Error.captureStackTrace(err, takeDemoFile);
    throw err;
  }

  return [demoPath, Buffer.from(demo), bodyDemofile] as const
}

function escapeName(currentTestName: string) {
  return Buffer.from(currentTestName)
    .map((c) => (charAllow.includes(c) ? c : 0x2d))
    .toString();
}

