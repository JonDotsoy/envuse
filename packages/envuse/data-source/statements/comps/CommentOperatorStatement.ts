import { Base } from "./Base";
import { BufferCursor } from "../lib/BufferCursor";
import { BCharType } from "../tdo/BCharType";
import { Space } from "./Space";
import { CharactersKey as K } from "../tdo/CharactersKey";
import { StatementObject, StatementObjectType } from "./StatementObject";
import { BaseSerializeOption } from "../tdo/BaseSerializeOption";

export type CommentOperatorStatementType = {
  $type: "CommentOperatorStatement";
  statements: StatementObjectType[];
};

export class CommentOperatorStatement extends Base {
  $type = "CommentOperatorStatement" as const;
  statements: StatementObject[] = [];

  prepare(bufferCursor: BufferCursor<BCharType>): void {
    while (true) {
      if (
        bufferCursor.current() === K.newLineLF ||
        bufferCursor.current() === K.numberSign ||
        bufferCursor.isClosed()
      ) {
        bufferCursor.forward();
        return;
      }

      this.statements.push(this.createElement(StatementObject));

      if (bufferCursor.current() === K.newLineLF) {
        bufferCursor.forward();
        return;
      }

      if (bufferCursor.current() === K.space) {
        this.createElement(Space);
      }
    }
  }

  toJSON() {
    return {
      ...super.toJSON(),
      statements: this.statements,
    };
  }

  static serialize(comp: CommentOperatorStatementType) {
    return Buffer.concat(
      comp.statements.map((s, i, { length }) =>
        Buffer.concat([
          StatementObject.serialize(s),
          Buffer.from(length - 1 === i ? "" : " "),
        ])
      )
    );
  }
}
