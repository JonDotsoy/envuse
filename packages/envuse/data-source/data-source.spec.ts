import { DataSource } from "./data-source";
import { takeDemoFile } from "./statements/lib/takeDemoFile";
import util from "util";
import { CommentOperator } from "./statements/comps/CommentOperator";
import { CommentOperatorStatement } from "./statements/comps/CommentOperatorStatement";
import { StatementObject } from "./statements/comps/StatementObject";
import { Base } from "./statements/comps/Base";
import { b } from "./statements/lib/toBuffer";

describe("DataSource", () => {
  it("shoud make a ast out", () => {
    const [filename1, body1] = takeDemoFile();

    const envuseFileParser = new DataSource(filename1, body1);
    expect(
      util.formatWithOptions(
        { depth: Infinity },
        "",
        envuseFileParser.toAstBody()
      )
    ).toMatchSnapshot();
  });

  it("shoud make a ast with operators", () => {
    const [filename, body] = takeDemoFile();

    const envuseFileParser = new DataSource(filename, body);
    expect(
      util.formatWithOptions(
        { depth: Infinity },
        "",
        envuseFileParser.toAstBody()
      )
    ).toMatchSnapshot();
    // console.log(util.formatWithOptions({ depth: Infinity }, '', envuseFileParser.toAstBody()))
  });

  it("shoud make a ast with multiple blocks", () => {
    const [filename, body] = takeDemoFile(".env");

    const envuseFileParser = new DataSource(filename, body);
    expect(
      util.formatWithOptions(
        { depth: Infinity },
        "",
        envuseFileParser.toAstBody()
      )
    ).toMatchSnapshot();
    // console.log(util.formatWithOptions({ depth: Infinity }, '', envuseFileParser.toAstBody()))
  });

  describe("parse statement", () => {
    it("parse statement operation", () => {
      const body = Buffer.from("true ===   false # asd");

      const cmp = Base.createElement(
        new CommentOperatorStatement("a", body, 0)
      );

      expect(cmp).toMatchSnapshot();
    });

    it("should parse statement with numbers", () => {
      const body = Buffer.from("34214324 ===   22_343.32 # comment");

      const cmp = Base.createElement(
        new CommentOperatorStatement("a", body, 0)
      );

      expect(cmp).toMatchSnapshot();
      // console.log(cmp)
    });

    it("should parse StatementObject string", () => {
      const [fl, body] = takeDemoFile();

      const cmp = Base.createElement(new StatementObject(fl, body, 0));

      // expect(cmp).toMatchSnapshot()
      // console.log(cmp)
    });

    it("should parse statement with string", () => {
      const [fl, body] = takeDemoFile();

      Base.createElement(new CommentOperatorStatement(fl, body, 0));

      // expect(cmp).toMatchSnapshot()
      // console.log(cmp)
    });

    it("should parse statement with string and special characters", () => {
      const [fl, body] = takeDemoFile(".special-characters.txt");

      const cmp = Base.createElement(new StatementObject(fl, body, 0));

      expect(cmp.value).toEqual("\n\t");
    });

    it("should parse instance path", () => {
      const body = Buffer.from("obj.a.b.VAL2");

      const cmp = Base.createElement(new StatementObject("a", body, 0));

      expect(cmp).toBeInstanceOf(StatementObject);
      expect(cmp.$type).toEqual("StatementObject");
      expect(cmp.type).toEqual("NameInstance");
      expect(cmp.value).toEqual(["obj", "a", "b", "VAL2"]);
    });

    it("should parse numbers", () => {
      const body = b("1_321.123");

      const cmp = Base.createElement(new StatementObject("a", body, 0));

      expect(cmp.value).toEqual(1_321.123);
    });

    it("should parse strings", () => {
      const body = b('"hi"');

      const cmp = Base.createElement(new StatementObject(null, body, 0));

      expect(cmp.value).toEqual("hi");
    });

    it("should parse statement with variables", () => {
      const body = Buffer.from("VAL1 ===   obj.a.b.VAL2 # comment");

      const cmp = Base.createElement(
        new CommentOperatorStatement("a", body, 0)
      );

      expect(cmp.elementList.map((e) => e.toString())).toMatchInlineSnapshot(`
        Array [
          "CommentOperatorStatement (0, 25): \\"VAL1 ===   obj.a.b.VAL2 #\\"",
          "StatementObject<NameInstance> (0, 4): \\"VAL1\\"",
          "Space (4, 5): \\" \\"",
          "StatementObject<StrictEqualitySymbol> (5, 8): \\"===\\"",
          "Space (8, 11): \\"   \\"",
          "StatementObject<NameInstance> (11, 23): \\"obj.a.b.VAL2\\"",
          "Space (23, 24): \\" \\"",
        ]
      `);
    });

    it("should parse end parsing comment code", () => {
      const body = b("#! if true\nFOO=bar");

      const cmp = Base.createElement(new CommentOperator("a", body, 0));

      // expect(cmp).toMatchInlineSnapshot(`Object {}`)
      expect(cmp.operator.raw.toString()).toEqual("if");
      // console.log(cmp)
    });

    it("should read property list children", () => {
      const [fl, body] = takeDemoFile();

      const envuseFileParser = new DataSource(fl, body);

      const b = envuseFileParser.toAstBody();

      for (const elem of b.elementList) {
        expect(elem).toBeInstanceOf(Base);
      }

      expect(
        b.elementList.map((e) => `${e.constructor.name} <${e.pos}, ${e.end}>`)
      ).toMatchInlineSnapshot(`
        Array [
          "Block <0, 35>",
          "Variable <0, 8>",
          "VariableKey <0, 3>",
          "SymbolEqual <3, 4>",
          "VariableValue <4, 8>",
          "CommentOperator <8, 34>",
          "Space <10, 11>",
          "VariableKey <11, 13>",
          "Space <13, 14>",
          "CommentOperatorStatement <14, 19>",
          "StatementObject <14, 18>",
          "Block <19, 34>",
          "SpaceNewLine <19, 20>",
          "Variable <20, 28>",
          "VariableKey <20, 23>",
          "SymbolEqual <23, 24>",
          "VariableValue <24, 28>",
          "SpaceNewLine <28, 29>",
          "CommentOperator <29, 34>",
          "Space <31, 32>",
          "VariableKey <32, 34>",
          "SpaceNewLine <34, 35>",
        ]
      `);
    });

    it("should parse comment operator correctly", () => {
      expect(
        DataSource.createDataSource({
          filename: null,
          body: Buffer.from(`#;if a.b.c.d ===\nfoo="bar"\n#;fi\n`),
        }).elementList.map((e) => e.toString())
      ).toMatchSnapshot();
      expect(
        DataSource.createDataSource({
          filename: null,
          body: Buffer.from(`#;if true\nfoo="bar"\n#;fi\n`),
        }).elementList.map((e) => e.toString())
      ).toMatchSnapshot();
      expect(
        DataSource.createDataSource({
          filename: null,
          body: Buffer.from(`#;if 123\nfoo="bar"\n#;fi\n`),
        }).elementList.map((e) => e.toString())
      ).toMatchSnapshot();
      expect(
        DataSource.createDataSource({
          filename: null,
          body: Buffer.from(`#;if var\nfoo="bar"\n#;fi\n`),
        }).elementList.map((e) => e.toString())
      ).toMatchSnapshot();
      expect(
        DataSource.createDataSource({
          filename: null,
          body: Buffer.from(`#;if foo === bar\nfoo="bar"\n#;fi\n`),
        }).elementList.map((e) => e.toString())
      ).toMatchSnapshot();
    });

    it("should parse block descriptive comment", () => {
      const [fl, demo] = takeDemoFile()

      const envuseFileParser = DataSource.createDataSource({ filename: fl, body: demo });

      expect(envuseFileParser).toMatchInlineSnapshot();
    });
  });
});

describe("DataSource file parse", () => {
  it("parse file", () => {
    const res = DataSource.parse({ body: b(`foo=var\nabc=def\naaa=32`) });

    expect(res.parsed).toEqual({
      foo: "var",
      abc: "def",
      aaa: "32",
    });
  });

  it("should parse with mixing conditionals", () => {
    const res1 = DataSource.parse({
      body: b(`foo=var\nabc=def\naaa=32\n#; if true\nbbb=ccc\n#; fi`),
    });

    expect(res1.parsed).toEqual({
      foo: "var",
      abc: "def",
      aaa: "32",
      bbb: "ccc",
    });

    const res2 = DataSource.parse({
      body: b(`foo=var\nabc=def\naaa=32\n#; if false\nbbb=ccc\n#; fi`),
    });

    expect(res2.parsed).toEqual({
      foo: "var",
      abc: "def",
      aaa: "32",
    });

    const res3 = DataSource.parse({
      body: b(
        `aaa=aaa\n#; if true\nbbb=bbb\n#; if false\nccc=ccc\n#; fi\n#; fi`
      ),
    });

    expect(res3.parsed).toEqual({
      aaa: "aaa",
      bbb: "bbb",
    });

    const res4 = DataSource.parse({
      body: b(
        `aaa=aaa\n#; if false\nbbb=bbb\n#; if true\nccc=ccc\n#; fi\n#; fi`
      ),
    });

    expect(res4.parsed).toEqual({
      aaa: "aaa",
    });
  });

  describe("complex sentences", () => {
    it("should parse with complex sentences (booleans)", () => {
      const res = DataSource.parse({
        body: b(`#; if true === true\naaa=aaa\n#; fi`),
      });

      expect(res.parsed).toStrictEqual({
        aaa: "aaa",
      });
    });

    it("should parse with complex sentences (booleans)", () => {
      const res = DataSource.parse({
        body: b(`bbb=bbb\n#; if true === true === false\naaa=aaa\n#; fi`),
      });

      expect(res.parsed).toStrictEqual({
        bbb: "bbb",
      });
    });

    it("should parse with complex sentences (strings)", () => {
      const res = DataSource.parse({
        body: b(`aaa=aaa\n#; if 'aaa' === 'aaa'\nbbb=bbb\n#; fi`),
      });

      expect(res.parsed).toStrictEqual({
        aaa: "aaa",
        bbb: "bbb",
      });
    });

    it("should parse with complex sentences (strings)", () => {
      const res = DataSource.parse({
        body: b(`aaa=aaa\n#; if 'ccc' === 'aaa'\nbbb=bbb\n#; fi`),
      });

      expect(res.parsed).toStrictEqual({
        aaa: "aaa",
      });
    });

    it("should parse with complex sentences (booleans)", () => {
      const res = DataSource.parse({
        body: b(`aaa=aaa\n#; if 12 === 12\nbbb=bbb\n#; fi`),
      });

      expect(res.parsed).toStrictEqual({
        aaa: "aaa",
        bbb: "bbb",
      });
    });

    it("should parse with complex sentences (booleans)", () => {
      const res = DataSource.parse({
        body: b(`aaa=aaa\n#; if 13_3 === 12\nbbb=bbb\n#; fi`),
      });

      expect(res.parsed).toStrictEqual({
        aaa: "aaa",
      });
    });

    it("should parse with complex sentences (name instance)", () => {
      const res = DataSource.parse({
        body: b(`aaa=aaa\n#; if 'aaa' === aaa\nbbb=bbb\n#; fi`),
      });

      expect(res.parsed).toStrictEqual({
        aaa: "aaa",
        bbb: "bbb",
      });
    });

    it("should parse with complex sentences (name instance)", () => {
      const res = DataSource.parse({
        body: b(`aaa=aaa\n#; if 'ccc' === aaa\nbbb=bbb\n#; fi`),
      });

      expect(res.parsed).toStrictEqual({
        aaa: "aaa",
      });
    });

    it("should parse with complex sentences (name instance and values)", () => {
      const res = DataSource.parse(
        { body: b(`aaa=aaa\n#; if 'ccc' === externalVal\nbbb=bbb\n#; fi`) },
        { externalVal: "ccc" }
      );

      expect(res.parsed).toStrictEqual({
        aaa: "aaa",
        bbb: "bbb",
      });
    });
  });
});

describe("DataSource file stringify", () => {
  it("should stringify ast", () => {
    const ast = DataSource.createDataSource(b(`foo=bar`));

    const str = JSON.stringify(ast, null, 2);

    expect(str).toMatchInlineSnapshot(`
      "{
        \\"$type\\": \\"Block\\",
        \\"pos\\": 0,
        \\"end\\": 7,
        \\"children\\": [
          {
            \\"$type\\": \\"Variable\\",
            \\"pos\\": 0,
            \\"end\\": 7,
            \\"keyVariable\\": {
              \\"$type\\": \\"VariableKey\\",
              \\"pos\\": 0,
              \\"end\\": 3,
              \\"value\\": \\"foo\\"
            },
            \\"valueVariable\\": {
              \\"$type\\": \\"VariableValue\\",
              \\"pos\\": 4,
              \\"end\\": 7,
              \\"value\\": \\"bar\\"
            }
          }
        ]
      }"
    `);
  });

  it("should stringify ast", () => {
    // const ast = EnvuseFileParser.parseToAst(b('a=b\n#; if true\nc=d\n#; fi'))
    const ast = DataSource.createDataSource(b("#; if true\n#; fi"));

    const str = JSON.stringify(ast, null, 2);

    expect(str).toMatchInlineSnapshot(`
      "{
        \\"$type\\": \\"Block\\",
        \\"pos\\": 0,
        \\"end\\": 16,
        \\"children\\": [
          {
            \\"$type\\": \\"CommentOperator\\",
            \\"pos\\": 0,
            \\"end\\": 16,
            \\"operator\\": {
              \\"$type\\": \\"VariableKey\\",
              \\"pos\\": 3,
              \\"end\\": 5,
              \\"value\\": \\"if\\"
            },
            \\"statement\\": {
              \\"$type\\": \\"CommentOperatorStatement\\",
              \\"pos\\": 6,
              \\"end\\": 11,
              \\"statements\\": [
                {
                  \\"$type\\": \\"StatementObject\\",
                  \\"pos\\": 6,
                  \\"end\\": 10,
                  \\"type\\": \\"Boolean\\",
                  \\"value\\": true
                }
              ]
            },
            \\"block\\": {
              \\"$type\\": \\"Block\\",
              \\"pos\\": 11,
              \\"end\\": 16,
              \\"children\\": [
                {
                  \\"$type\\": \\"CommentOperator\\",
                  \\"pos\\": 11,
                  \\"end\\": 16,
                  \\"operator\\": {
                    \\"$type\\": \\"VariableKey\\",
                    \\"pos\\": 14,
                    \\"end\\": 16,
                    \\"value\\": \\"fi\\"
                  }
                }
              ]
            }
          }
        ]
      }"
    `);
  });
});
