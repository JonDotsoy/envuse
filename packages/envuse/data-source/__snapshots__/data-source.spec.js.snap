// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSource file stringify should stringify ast 1`] = `
"{
  \\"$type\\": \\"Block\\",
  \\"pos\\": 0,
  \\"end\\": 8,
  \\"children\\": [
    {
      \\"$type\\": \\"Variable\\",
      \\"pos\\": 0,
      \\"end\\": 8,
      \\"keyVariable\\": {
        \\"$type\\": \\"VariableKey\\",
        \\"pos\\": 0,
        \\"end\\": 3,
        \\"value\\": \\"foo\\"
      },
      \\"valueVariable\\": {
        \\"$type\\": \\"VariableValue\\",
        \\"pos\\": 4,
        \\"end\\": 7,
        \\"value\\": \\"bar\\"
      }
    }
  ]
}"
`;

exports[`DataSource file stringify should stringify ast 2`] = `
"{
  \\"$type\\": \\"Block\\",
  \\"pos\\": 0,
  \\"end\\": 16,
  \\"children\\": [
    {
      \\"$type\\": \\"CommentOperator\\",
      \\"pos\\": 0,
      \\"end\\": 16,
      \\"operator\\": {
        \\"$type\\": \\"VariableKey\\",
        \\"pos\\": 3,
        \\"end\\": 5,
        \\"value\\": \\"if\\"
      },
      \\"statement\\": {
        \\"$type\\": \\"CommentOperatorStatement\\",
        \\"pos\\": 6,
        \\"end\\": 11,
        \\"statements\\": [
          {
            \\"$type\\": \\"StatementObject\\",
            \\"pos\\": 6,
            \\"end\\": 10,
            \\"type\\": \\"Boolean\\",
            \\"value\\": true
          }
        ]
      },
      \\"block\\": {
        \\"$type\\": \\"Block\\",
        \\"pos\\": 11,
        \\"end\\": 16,
        \\"children\\": [
          {
            \\"$type\\": \\"CommentOperator\\",
            \\"pos\\": 11,
            \\"end\\": 16,
            \\"operator\\": {
              \\"$type\\": \\"VariableKey\\",
              \\"pos\\": 14,
              \\"end\\": 16,
              \\"value\\": \\"fi\\"
            }
          }
        ]
      }
    }
  ]
}"
`;

exports[`DataSource parse statement parse statement operation 1`] = `
Object {
  "$type": "CommentOperatorStatement",
  "children": undefined,
  "end": 18,
  "pos": 0,
  "statements": Array [
    Object {
      "$type": "StatementObject",
      "children": undefined,
      "end": 4,
      "pos": 0,
      "type": "Boolean",
      "value": true,
    },
    Object {
      "$type": "StatementObject",
      "children": undefined,
      "end": 8,
      "pos": 5,
      "type": "StrictEqualitySymbol",
      "value": undefined,
    },
    Object {
      "$type": "StatementObject",
      "children": undefined,
      "end": 16,
      "pos": 11,
      "type": "Boolean",
      "value": false,
    },
  ],
}
`;

exports[`DataSource parse statement should parse comment operator correctly 1`] = `
Array [
  "Block (0, 32): \\"#;if a.b....",
  "CommentOperator (0, 31): \\"#;if a.b....",
  "VariableKey (2, 4): \\"if\\"",
  "Space (4, 5): \\" \\"",
  "CommentOperatorStatement (5, 17): \\"a.b.c.d =...",
  "StatementObject<NameInstance> (5, 12): \\"a.b.c.d\\"",
  "Space (12, 13): \\" \\"",
  "StatementObject<StrictEqualitySymbol> (13, 16): \\"===\\"",
  "Block (17, 31): \\"foo=\\\\\\"bar...",
  "Variable (17, 27): \\"foo=\\\\\\"bar...",
  "VariableKey (17, 20): \\"foo\\"",
  "SymbolEqual (20, 21): \\"=\\"",
  "VariableValue (21, 26): \\"\\\\\\"bar\\\\\\"\\"",
  "CommentOperator (27, 31): \\"#;fi\\"",
  "VariableKey (29, 31): \\"fi\\"",
  "SpaceNewLine (31, 32): \\"\\\\n\\"",
]
`;

exports[`DataSource parse statement should parse comment operator correctly 2`] = `
Array [
  "Block (0, 25): \\"#;if true...",
  "CommentOperator (0, 24): \\"#;if true...",
  "VariableKey (2, 4): \\"if\\"",
  "Space (4, 5): \\" \\"",
  "CommentOperatorStatement (5, 10): \\"true\\\\n\\"",
  "StatementObject<Boolean> (5, 9): \\"true\\"",
  "Block (10, 24): \\"foo=\\\\\\"bar...",
  "Variable (10, 20): \\"foo=\\\\\\"bar...",
  "VariableKey (10, 13): \\"foo\\"",
  "SymbolEqual (13, 14): \\"=\\"",
  "VariableValue (14, 19): \\"\\\\\\"bar\\\\\\"\\"",
  "CommentOperator (20, 24): \\"#;fi\\"",
  "VariableKey (22, 24): \\"fi\\"",
  "SpaceNewLine (24, 25): \\"\\\\n\\"",
]
`;

exports[`DataSource parse statement should parse comment operator correctly 3`] = `
Array [
  "Block (0, 24): \\"#;if 123\\\\...",
  "CommentOperator (0, 23): \\"#;if 123\\\\...",
  "VariableKey (2, 4): \\"if\\"",
  "Space (4, 5): \\" \\"",
  "CommentOperatorStatement (5, 9): \\"123\\\\n\\"",
  "StatementObject<Number> (5, 8): \\"123\\"",
  "Block (9, 23): \\"foo=\\\\\\"bar...",
  "Variable (9, 19): \\"foo=\\\\\\"bar...",
  "VariableKey (9, 12): \\"foo\\"",
  "SymbolEqual (12, 13): \\"=\\"",
  "VariableValue (13, 18): \\"\\\\\\"bar\\\\\\"\\"",
  "CommentOperator (19, 23): \\"#;fi\\"",
  "VariableKey (21, 23): \\"fi\\"",
  "SpaceNewLine (23, 24): \\"\\\\n\\"",
]
`;

exports[`DataSource parse statement should parse comment operator correctly 4`] = `
Array [
  "Block (0, 24): \\"#;if var\\\\...",
  "CommentOperator (0, 23): \\"#;if var\\\\...",
  "VariableKey (2, 4): \\"if\\"",
  "Space (4, 5): \\" \\"",
  "CommentOperatorStatement (5, 9): \\"var\\\\n\\"",
  "StatementObject<NameInstance> (5, 8): \\"var\\"",
  "Block (9, 23): \\"foo=\\\\\\"bar...",
  "Variable (9, 19): \\"foo=\\\\\\"bar...",
  "VariableKey (9, 12): \\"foo\\"",
  "SymbolEqual (12, 13): \\"=\\"",
  "VariableValue (13, 18): \\"\\\\\\"bar\\\\\\"\\"",
  "CommentOperator (19, 23): \\"#;fi\\"",
  "VariableKey (21, 23): \\"fi\\"",
  "SpaceNewLine (23, 24): \\"\\\\n\\"",
]
`;

exports[`DataSource parse statement should parse comment operator correctly 5`] = `
Array [
  "Block (0, 32): \\"#;if foo ...",
  "CommentOperator (0, 31): \\"#;if foo ...",
  "VariableKey (2, 4): \\"if\\"",
  "Space (4, 5): \\" \\"",
  "CommentOperatorStatement (5, 17): \\"foo === b...",
  "StatementObject<NameInstance> (5, 8): \\"foo\\"",
  "Space (8, 9): \\" \\"",
  "StatementObject<StrictEqualitySymbol> (9, 12): \\"===\\"",
  "Space (12, 13): \\" \\"",
  "StatementObject<NameInstance> (13, 16): \\"bar\\"",
  "Block (17, 31): \\"foo=\\\\\\"bar...",
  "Variable (17, 27): \\"foo=\\\\\\"bar...",
  "VariableKey (17, 20): \\"foo\\"",
  "SymbolEqual (20, 21): \\"=\\"",
  "VariableValue (21, 26): \\"\\\\\\"bar\\\\\\"\\"",
  "CommentOperator (27, 31): \\"#;fi\\"",
  "VariableKey (29, 31): \\"fi\\"",
  "SpaceNewLine (31, 32): \\"\\\\n\\"",
]
`;

exports[`DataSource parse statement should parse statement with numbers 1`] = `
Object {
  "$type": "CommentOperatorStatement",
  "children": undefined,
  "end": 26,
  "pos": 0,
  "statements": Array [
    Object {
      "$type": "StatementObject",
      "children": undefined,
      "end": 8,
      "pos": 0,
      "type": "Number",
      "value": 34214324,
    },
    Object {
      "$type": "StatementObject",
      "children": undefined,
      "end": 12,
      "pos": 9,
      "type": "StrictEqualitySymbol",
      "value": undefined,
    },
    Object {
      "$type": "StatementObject",
      "children": undefined,
      "end": 24,
      "pos": 15,
      "type": "Number",
      "value": 22343.32,
    },
  ],
}
`;

exports[`DataSource parse statement should parse statement with variables 1`] = `
Array [
  "CommentOperatorStatement (0, 25): \\"VAL1 === ...",
  "StatementObject<NameInstance> (0, 4): \\"VAL1\\"",
  "Space (4, 5): \\" \\"",
  "StatementObject<StrictEqualitySymbol> (5, 8): \\"===\\"",
  "Space (8, 11): \\"   \\"",
  "StatementObject<NameInstance> (11, 23): \\"obj.a.b.VAL2\\"",
  "Space (23, 24): \\" \\"",
]
`;

exports[`DataSource parse statement should parse variable with comment 1`] = `
"[
  Block (0, 30): \\"FOO : num...
    Variable (0, 30): \\"FOO : num...
      ...,
  Variable (0, 30): \\"FOO : num...
    VariableKey (0, 3): \\"FOO\\"
    SymbolColon (4, 5): \\":\\"
    VariableKey (6, 12): \\"number\\"
    SymbolEqual (13, 14): \\"=\\"
    VariableValue (15, 20): \\"\\\\\\"123\\\\\\"\\"
    CommentInline (21, 30): \\"# comment...,
  VariableKey (0, 3): \\"FOO\\",
  Space (3, 4): \\" \\",
  SymbolColon (4, 5): \\":\\",
  Space (5, 6): \\" \\",
  VariableKey (6, 12): \\"number\\",
  Space (12, 13): \\" \\",
  SymbolEqual (13, 14): \\"=\\",
  Space (14, 15): \\" \\",
  VariableValue (15, 20): \\"\\\\\\"123\\\\\\"\\",
  Space (20, 21): \\" \\",
  CommentInline (21, 30): \\"# comment...,
  Space (22, 23): \\" \\"
]"
`;

exports[`DataSource parse statement should parse variable with comment 2`] = `
"[
  Block (0, 31): \\"FOO : num...
    Variable (0, 31): \\"FOO : num...
      ...,
  Variable (0, 31): \\"FOO : num...
    VariableKey (0, 3): \\"FOO\\"
    SymbolColon (4, 5): \\":\\"
    VariableKey (6, 12): \\"number\\"
    SymbolEqual (13, 14): \\"=\\"
    VariableValue (15, 21): \\"'12#3'\\"
    CommentInline (22, 31): \\"# comment...,
  VariableKey (0, 3): \\"FOO\\",
  Space (3, 4): \\" \\",
  SymbolColon (4, 5): \\":\\",
  Space (5, 6): \\" \\",
  VariableKey (6, 12): \\"number\\",
  Space (12, 13): \\" \\",
  SymbolEqual (13, 14): \\"=\\",
  Space (14, 15): \\" \\",
  VariableValue (15, 21): \\"'12#3'\\",
  Space (21, 22): \\" \\",
  CommentInline (22, 31): \\"# comment...,
  Space (23, 24): \\" \\"
]"
`;

exports[`DataSource parse statement should parse variable with comment 3`] = `
"Block (0, 35): \\"FOO : num...
  Variable (0, 35): \\"FOO : num...
    VariableKey (0, 3): \\"FOO\\"
    SymbolColon (4, 5): \\":\\"
    VariableKey (6, 12): \\"number\\"
    SymbolEqual (13, 14): \\"=\\"
    VariableValue (15, 22): \\"HOLA HI\\"
    CommentInline (26, 35): \\"# comment..."
`;

exports[`DataSource parse statement should parse variable with comment 4`] = `
"Block (0, 45): \\"FOO : num...
  Variable (0, 36): \\"FOO : num...
    VariableKey (0, 3): \\"FOO\\"
    SymbolColon (4, 5): \\":\\"
    VariableKey (6, 12): \\"number\\"
    SymbolEqual (13, 14): \\"=\\"
    VariableValue (15, 22): \\"HOLA HI\\"
    CommentInline (26, 36): \\"# comment...
  Variable (36, 45): \\"A='B#'#sa...
    VariableKey (36, 37): \\"A\\"
    SymbolEqual (37, 38): \\"=\\"
    VariableValue (38, 42): \\"'B#'\\"
    CommentInline (42, 45): \\"#sa\\""
`;

exports[`DataSource parse statement should parse variable with type 1`] = `
"[
  Block (0, 19): \\"FOO : num...
    Variable (0, 19): \\"FOO : num...
      ...,
  Variable (0, 19): \\"FOO : num...
    VariableKey (0, 3): \\"FOO\\"
    SymbolColon (4, 5): \\":\\"
    VariableKey (6, 12): \\"number\\"
    SymbolEqual (13, 14): \\"=\\"
    VariableValue (15, 18): \\"123\\",
  VariableKey (0, 3): \\"FOO\\",
  Space (3, 4): \\" \\",
  SymbolColon (4, 5): \\":\\",
  Space (5, 6): \\" \\",
  VariableKey (6, 12): \\"number\\",
  Space (12, 13): \\" \\",
  SymbolEqual (13, 14): \\"=\\",
  Space (14, 15): \\" \\",
  VariableValue (15, 18): \\"123\\"
]"
`;

exports[`DataSource should parse variable with backslash 1`] = `
"Block (0, 7): \\"A='\\\\\\\\''\\\\n...
  Variable (0, 7): \\"A='\\\\\\\\''\\\\n...
    VariableKey (0, 1): \\"A\\"
    SymbolEqual (1, 2): \\"=\\"
    VariableValue (2, 6): \\"'\\\\\\\\''\\""
`;

exports[`DataSource should parse variable with backslash 2`] = `
"Block (0, 7): \\"A=\\\\\\"\\\\\\\\\\\\\\"\\\\...
  Variable (0, 7): \\"A=\\\\\\"\\\\\\\\\\\\\\"\\\\...
    VariableKey (0, 1): \\"A\\"
    SymbolEqual (1, 2): \\"=\\"
    VariableValue (2, 6): \\"\\\\\\"\\\\\\\\\\\\\\"\\\\\\"\\""
`;
